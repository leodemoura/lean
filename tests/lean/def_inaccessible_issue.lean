open nat

set_option pp.binder_types true

inductive bv : nat → Type
| nil  : bv 0
| cons : ∀ (n) (hd : bool) (tl : bv n), bv (succ n)

open bv

variable (f : bool → bool → bool)

set_option trace.type_context.is_def_eq true
set_option trace.type_context.is_def_eq_detail true
set_option trace.type_context.eqn_lemmas true
set_option pp.numerals false
set_option pp.implicit true
set_option pp.purify_metavars true

definition map2 : ∀ {n}, bv n → bv n → bv n
| 0     nil            nil             := nil
| (n+1) (cons .n b1 v1) (cons .n b2 v2) := cons n (f b1 b2) (map2 v1 v2)
exit

check map2._main.equations.eqn_2

/- defining function again without simplifying the type of automatically generated lemmas -/

definition map2' : ∀ {n}, bv n → bv n → bv n
| 0     nil            nil             := nil
| (n+1) (cons .n b1 v1) (cons .n b2 v2) := cons n (f b1 b2) (map2' v1 v2)

check map2'._main.equations.eqn_2
