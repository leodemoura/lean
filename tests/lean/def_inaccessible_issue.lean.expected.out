map2._main.equations.eqn_2 :
  ∀ (f : bool → bool → bool) (n : ℕ) (b1 : bool)
  (v1 :
    bv
      (prod.fst
         (prod.fst
            (nat.rec
              (λ (a : ℕ),
                nat.cases_on 0 (λ (_F : nat.below 0), a)
                  (λ (a_1 : ℕ) (_F : nat.below (succ a_1)), succ (prod.fst (prod.fst _F) a))
                  poly_unit.star, poly_unit.star)
              (λ (a : ℕ) (ih_1 : (ℕ → ℕ) × nat.below a),
                 (λ (a_1 : ℕ),
                   nat.cases_on (succ a) (λ (_F : nat.below 0), a_1)
                     (λ (a : ℕ) (_F : nat.below (succ a)), succ (prod.fst (prod.fst _F) a_1))
                     (ih_1, poly_unit.star), ih_1, poly_unit.star))
              0, poly_unit.star))
         n)) (b2 : bool)
  (v2 :
    bv
      (prod.fst
         (prod.fst
            (nat.rec
              (λ (a : ℕ),
                nat.cases_on 0 (λ (_F : nat.below 0), a)
                  (λ (a_1 : ℕ) (_F : nat.below (succ a_1)), succ (prod.fst (prod.fst _F) a))
                  poly_unit.star, poly_unit.star)
              (λ (a : ℕ) (ih_1 : (ℕ → ℕ) × nat.below a),
                 (λ (a_1 : ℕ),
                   nat.cases_on (succ a) (λ (_F : nat.below 0), a_1)
                     (λ (a : ℕ) (_F : nat.below (succ a)), succ (prod.fst (prod.fst _F) a_1))
                     (ih_1, poly_unit.star), ih_1, poly_unit.star))
              0, poly_unit.star))
         n)),
    map2._main f
      (cons
         (prod.fst
            (prod.fst
               (nat.rec
                 (λ (a : ℕ),
                   nat.cases_on 0 (λ (_F : nat.below 0), a)
                     (λ (a_1 : ℕ) (_F : nat.below (succ a_1)), succ (prod.fst (prod.fst _F) a))
                     poly_unit.star, poly_unit.star)
                 (λ (a : ℕ) (ih_1 : (ℕ → ℕ) × nat.below a),
                    (λ (a_1 : ℕ),
                      nat.cases_on (succ a) (λ (_F : nat.below 0), a_1)
                        (λ (a : ℕ) (_F : nat.below (succ a)), succ (prod.fst (prod.fst _F) a_1))
                        (ih_1, poly_unit.star), ih_1, poly_unit.star))
                 0, poly_unit.star))
            n)
         b1
         v1)
      (cons
         (prod.fst
            (prod.fst
               (nat.rec
                 (λ (a : ℕ),
                   nat.cases_on 0 (λ (_F : nat.below 0), a)
                     (λ (a_1 : ℕ) (_F : nat.below (succ a_1)), succ (prod.fst (prod.fst _F) a))
                     poly_unit.star, poly_unit.star)
                 (λ (a : ℕ) (ih_1 : (ℕ → ℕ) × nat.below a),
                    (λ (a_1 : ℕ),
                      nat.cases_on (succ a) (λ (_F : nat.below 0), a_1)
                        (λ (a : ℕ) (_F : nat.below (succ a)), succ (prod.fst (prod.fst _F) a_1))
                        (ih_1, poly_unit.star), ih_1, poly_unit.star))
                 0, poly_unit.star))
            n)
         b2
         v2) = cons n (f b1 b2) (map2._main f v1 v2)
map2'._main.equations.eqn_2 :
  ∀ (f : bool → bool → bool) (n : ℕ) (b1 : bool)
  (v1 :
    bv
      (prod.fst
         (prod.fst
            (nat.rec
              (λ (a : ℕ),
                nat.cases_on 0 (λ (_F : nat.below 0), a)
                  (λ (a_1 : ℕ) (_F : nat.below (succ a_1)), succ (prod.fst (prod.fst _F) a))
                  poly_unit.star, poly_unit.star)
              (λ (a : ℕ) (ih_1 : (ℕ → ℕ) × nat.below a),
                 (λ (a_1 : ℕ),
                   nat.cases_on (succ a) (λ (_F : nat.below 0), a_1)
                     (λ (a : ℕ) (_F : nat.below (succ a)), succ (prod.fst (prod.fst _F) a_1))
                     (ih_1, poly_unit.star), ih_1, poly_unit.star))
              0, poly_unit.star))
         n)) (b2 : bool)
  (v2 :
    bv
      (prod.fst
         (prod.fst
            (nat.rec
              (λ (a : ℕ),
                nat.cases_on 0 (λ (_F : nat.below 0), a)
                  (λ (a_1 : ℕ) (_F : nat.below (succ a_1)), succ (prod.fst (prod.fst _F) a))
                  poly_unit.star, poly_unit.star)
              (λ (a : ℕ) (ih_1 : (ℕ → ℕ) × nat.below a),
                 (λ (a_1 : ℕ),
                   nat.cases_on (succ a) (λ (_F : nat.below 0), a_1)
                     (λ (a : ℕ) (_F : nat.below (succ a)), succ (prod.fst (prod.fst _F) a_1))
                     (ih_1, poly_unit.star), ih_1, poly_unit.star))
              0, poly_unit.star))
         n)),
    map2'._main f
      (cons
         (prod.fst
            (prod.fst
               (nat.rec
                 (λ (a : ℕ),
                   nat.cases_on 0 (λ (_F : nat.below 0), a)
                     (λ (a_1 : ℕ) (_F : nat.below (succ a_1)), succ (prod.fst (prod.fst _F) a))
                     poly_unit.star, poly_unit.star)
                 (λ (a : ℕ) (ih_1 : (ℕ → ℕ) × nat.below a),
                    (λ (a_1 : ℕ),
                      nat.cases_on (succ a) (λ (_F : nat.below 0), a_1)
                        (λ (a : ℕ) (_F : nat.below (succ a)), succ (prod.fst (prod.fst _F) a_1))
                        (ih_1, poly_unit.star), ih_1, poly_unit.star))
                 0, poly_unit.star))
            n)
         b1
         v1)
      (cons
         (prod.fst
            (prod.fst
               (nat.rec
                 (λ (a : ℕ),
                   nat.cases_on 0 (λ (_F : nat.below 0), a)
                     (λ (a_1 : ℕ) (_F : nat.below (succ a_1)), succ (prod.fst (prod.fst _F) a))
                     poly_unit.star, poly_unit.star)
                 (λ (a : ℕ) (ih_1 : (ℕ → ℕ) × nat.below a),
                    (λ (a_1 : ℕ),
                      nat.cases_on (succ a) (λ (_F : nat.below 0), a_1)
                        (λ (a : ℕ) (_F : nat.below (succ a)), succ (prod.fst (prod.fst _F) a_1))
                        (ih_1, poly_unit.star), ih_1, poly_unit.star))
                 0, poly_unit.star))
            n)
         b2
         v2) = cons n (f b1 b2) (map2'._main f v1 v2)
